### PR TITLE
Implement ReplaceNidsStep

### LINKED ISSUE #1376: replace WriteConfStep with `lctl replace_nids`
Replace the `WriteConfStep` with `lctl replace_nids` in Rust, and then delete the corresponding action plugin writeconf_target.

### PR DESCRIPTION
**Pull request by mkpankov**
Signed-off-by: Michael Pankov <work@michaelpankov.com>

Fixes #1376

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1416)

### CODE REVIEW

**Comment by mkpankov** [comment]
Completely untested ATM as I don't know how to run Python tests. Would appreciate pointers.

Also I'm confused as to what `target` is vs. `path`.

**Comment by jgrund** [comment]
To run unit / service / behave tests: 

https://whamcloud.github.io/Online-Help/docs/Contributor_Docs/cd_integrated-manager-for-lustre_Unit_Testing.html

**Comment by mkpankov** [comment]
Current status: blocked by having no tests, and current testing model doesn't start `iml-action-runner` or mock it, so it's simply down when test tries to call Rust agent.

Need to figure that part out.

**Comment by mkpankov** [comment]
Unit test pass locally now

**Comment by mkpankov** [comment]
Ready for review, I think.

File: tests/unit/services/job_scheduler/host/test_host.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
 from copy import deepcopy
 from itertools import chain
+from mock import call, MagicMock, patch
 import json
+import mock
```
Comment (ip1981) [inline tests/unit/services/job_scheduler/host/test_host.py]: It looks like if you add `patch` to the above import, you will not need this line.

File: chroma_core/models/host.py [hunk 2]
```diff
@@ -1578,22 +1578,13 @@ class Meta:
...
-            agent_args["mgsnode"] = args["mgsnode"]
-            agent_args["writeconf"] = True
-
-        fail_nids = args["fail_nids"]
-        if fail_nids:
-            agent_args["failnode"] = fail_nids
-        self.invoke_agent(args["host"], "writeconf_target", agent_args)
+        device = args["path"]
+        nids = args["nids"]
+        flattened_nids = [e for t in nids for e in t]
```
Comment (ip1981) [inline chroma_core/models/host.py]: Technically, `flattened_nids` and `nids` may include duplicate elements. Is it ok?

File: tests/unit/services/job_scheduler/host/test_host.py (hunk 1 above)
Comment (mkpankov) [reply tests/unit/services/job_scheduler/host/test_host.py]: Done

File: chroma_core/models/host.py (hunk 2 above)
Comment (mkpankov) [reply chroma_core/models/host.py]: I think that shouldn't be the case based on what I've seen. But it's not handled here, just as in previous implementation.

**Comment by jgrund** [comment]
@mkpankov You can test this by changing nids on the IML server detail page.

**Comment by mkpankov** [comment]
@jgrund I tried doing that, it seems it doesn't trigger the `UpdateNidsJob`, as steps are different

Screenshot from 2020-01-09 10-04-42

Looking at places calling that job, I could only find [`NidsHandler`](https://github.com/whamcloud/integrated-manager-for-lustre/blob/master/chroma_cli/handlers.py#L581). Will try testing using that, somehow.

**Comment by mkpankov** [comment]
I've been able to run the CLI once, and the job itself has failed because the FS wasn't mounted.

Now I mounted the FS and try to rerun the job with
```
sudo PYTHONPATH=$PWD python chroma_cli/main.py --username admin --password lustre nid update
```

It fails with messages in `/var/log/chroma/job_scheduler.log`:
```
[2020-01-09 07:44:24,906: INFO/command_plan] add_jobs: created Job 575 (Update NIDs on 4 hosts)
[2020-01-09 07:44:25,740: ERROR/rpc] RunOneRpc: exception calling run_jobs: Traceback (most recent call last):

  File "/usr/share/chroma-manager/chroma_core/services/rpc.py", line 107, in run
    "result": self.rpc._local_call(self.body["method"], *self.body["args"], **self.body["kwargs"]),

  File "/usr/share/chroma-manager/chroma_core/services/rpc.py", line 622, in _local_call
    return fn(*args, **kwargs)

  File "/usr/share/chroma-manager/chroma_core/services/job_scheduler/job_scheduler.py", line 893, in run_jobs
    result = self.CommandPlan.command_run_jobs(job_dicts, message)

  File "/usr/share/chroma-manager/chroma_core/services/job_scheduler/command_plan.py", line 512, in command_run_jobs
    self.add_jobs(jobs, command, job_deps_map)

  File "/usr/lib/python2.7/site-packages/django/db/transaction.py", line 223, in __exit__
    connection.commit()

  File "/usr/lib/python2.7/site-packages/django/db/backends/base/base.py", line 262, in commit
    self._commit()

  File "/usr/lib/python2.7/site-packages/django/db/backends/base/base.py", line 236, in _commit
    return self.connection.commit()

  File "/usr/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)

  File "/usr/lib/python2.7/site-packages/django/db/backends/base/base.py", line 236, in _commit
    return self.connection.commit()

IntegrityError: insert or update on table "chroma_core_command_jobs" violates foreign key constraint "chroma_core_command_jobs_job_id_9bef0177_fk_chroma_core_job_id"
DETAIL:  Key (job_id)=(520) is not present in table "chroma_core_job".
```

@jgrund  How do I un-break my database w/o complete reinstall?

**Comment by mkpankov** [comment]
I've updated code to do the correct thing, but I think I'll polish it a little.

The current version works:
Screenshot from 2020-01-10 16-18-17
Screenshot from 2020-01-10 16-18-19
Screenshot from 2020-01-10 16-18-29

And fs starts successfully afterwards.

File: chroma_core/models/host.py [hunk 3]
```diff
@@ -1647,17 +1649,25 @@ def get_deps(self):
...
+        from chroma_core.models.target import ManagedMgs
+
         locks = []
         filesystems, targets = self._targets_on_hosts()
 
         for target in targets:
-            locks.append(
-                StateLock(job=self, locked_item=target, begin_state="unmounted", end_state="unmounted", write=True)
-            )
+            if not issubclass(target.downcast_class, ManagedMgs):
```
Comment (ip1981) [inline chroma_core/models/host.py]: nit:  could remove `not` and swap branches.

File: tests/unit/services/job_scheduler/host/test_host.py [hunk 4]
```diff
@@ -79,7 +80,10 @@ class TestUpdateNids(NidTestCase):
         },
     }
 
-    def test_mgs_nid_change(self):
+    @patch("chroma_core.lib.job.Step.invoke_rust_agent", return_value=MagicMock())
+    def test_mgs_nid_change(self, invoke):
+        invoke.return_value = """{"Ok": ""}"""
```
Comment (ip1981) [inline tests/unit/services/job_scheduler/host/test_host.py]:
```suggestion
        invoke.return_value = '{"Ok": ""}'
```

Comment (mkpankov) [reply tests/unit/services/job_scheduler/host/test_host.py]: Done

File: chroma_core/models/host.py (hunk 3 above)
Comment (mkpankov) [reply chroma_core/models/host.py]: Done

File: chroma_core/models/host.py [hunk 5]
```diff
@@ -1683,20 +1684,17 @@ def get_steps(self):
             target = target.downcast()
             primary_tm = target.managedtargetmount_set.get(primary=True)
             steps.append((MountOrImportStep, MountOrImportStep.create_parameters(target, primary_tm.host, False)))
```
Comment (utopiabound) [inline chroma_core/models/host.py]: Is this step needed?  Ensuring ZPOOLs were imported was important for writeconf, but not for replace nids.

Comment (mkpankov) [reply chroma_core/models/host.py]: Seems to work fine w/o it, removing

**Comment by mkpankov** [comment]
@jgrund sure thing

**Comment by jgrund** [comment]
@mkpankov Did line endings change in that file, looks like the whole file is being highlighted for diff.

**Comment by jgrund** [comment]
They were CRLF, repushed.

**Pull request merged**